Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift

```diff
@@ -0,0 +1,547 @@
+            merchantCode: (self.config.options as! MerchantOptions).merchantId,
+            refNo: self.primerTransactionId,
+            prodDesc: PrimerAPIConfiguration.current!.clientSession!.order!.lineItems!.compactMap({ $0.description }).joined(separator: ", "),
+            userName: "\(PrimerAPIConfiguration.current!.clientSession!.customer!.firstName!) \(PrimerAPIConfiguration.current!.clientSession!.customer!.lastName!)",
+            userEmail: PrimerAPIConfiguration.current!.clientSession!.customer!.emailAddress!,
+            userContact: PrimerAPIConfiguration.current!.clientSession!.customer!.mobileNumber!,
+            country: "MY",
+            backendPostURL: self.backendCallbackUrl!.absoluteString,
+            remark: nil,
+            lang: "UTF-8")
+    }
+
+    
+    override func presentPaymentMethodUserInterface() -> Promise<Void> {
+        return Promise { seal in
+            DispatchQueue.main.async { [unowned self] in
+                self.primerIPay88ViewController = PrimerIPay88ViewController(delegate: self, payment: self.primerIPay88Payment!)
+
+                if #available(iOS 13.0, *) {
+                    self.primerIPay88ViewController.isModalInPresentation = true
+                    self.primerIPay88ViewController.modalPresentationStyle = .fullScreen
+                }
+
+                self.willPresentPaymentMethodUI?()
+                PrimerUIManager.primerRootViewController?.present(self.primerIPay88ViewController, animated: true, completion: {
+                    DispatchQueue.main.async {
+                        PrimerHeadlessUniversalCheckout.current.delegate?.primerHeadlessUniversalCheckoutPaymentMethodDidShow?(for: self.config.type)
+                        self.didPresentPaymentMethodUI?()
+                        seal.fulfill(())
+                    }
+                })
+                
+                self.didComplete = { [unowned self] in
+                    DispatchQueue.main.async { [unowned self] in
+                        PrimerUIManager.primerRootViewController?.showLoadingScreenIfNeeded(imageView: nil, message: nil)
+                        self.primerIPay88ViewController?.dismiss(animated: true, completion: {
+                            
+                        })
+                    }
+                }
+            }
+        }
+    }
+    
+    override func awaitUserInput() -> Promise<Void> {
+        return Promise { seal in
+            let pollingModule = PollingModule(url: self.statusUrl)
+            self.didCancel = {
+                let err = PrimerError.cancelled(
+                    paymentMethodType: self.config.type,
+                    userInfo: ["file": #file, "class": "\(Self.self)", "function": #function, "line": "\(#line)"],
+                    diagnosticsId: nil)
+                ErrorHandler.handle(error: err)
+                pollingModule.cancel(withError: err)
+            }
+            
+            self.didFail = { err in
```

Comment:
If I get this right: this automatically outputs a log to Primer?

Reply:
If it cancels on HUC, an `cancellation` error will be returned. If it's dropin, it will return on the dropin UI.

If it's an error, it will be passed via the delegates.